feat(core): add undecorated-base-class migration schematic

devversion opened the pull request:
Introduces a new upgrade schematic that automatically decorates
`@Directive` / `@Component` base classes that use dependency injection
with the `@Directive` decorator. The new dummy directives are also
added to the corresponding "NgModule" declarations in order to not
cause any compilation warnings.

Adding the `@Directive` decorator ensures that "ngtsc" can properly
create the needed Ivy fields so that dependency injection works
as expected once Ivy is enabled in the application.

References FW-1238

trotyl commented on packages/core/schematics/migrations/undecorated-base-class/index.ts:
```diff
@@ -0,0 +1,128 @@
+            directiveExpr, undefined,
+            [ts.createObjectLiteral(
+                [ts.createPropertyAssignment(
+                    'selector', ts.createStringLiteral(`_base_class_${selectorIdx++}`))],
```
Why is the magic string in user source code? Also, the current convention should be adding parameter-less `@Injectable()` for base classes needed for DI, this is why parameter-less `@Injectable()` designed to be no-op and can be combined with other decorators: https://github.com/angular/angular/blob/0df719a4614dbe60c3bc6375a63ca95e340babba/packages/compiler-cli/test/ngtsc/ngtsc_spec.ts#L1021-L1035

devversion replied:
We need to use `@Directive` or `@Component` so that `ngtsc` can generate the proper `inject` call expressions. Since the base-classes can also inject from the directive injector, `ngtsc` needs to be able to generate the `directiveInject(..)` calls in order to use the proper injector.

I agree with the "magic string" confusion though. I brought that up as well, but it's not quite clear what we will do. We can possibly loosen the type for `@Directive`.

trotyl replied:
@devversion Would it possible to switch injection scope at factory? Like:

```typescript
function defineComponent(def: ComponentDef) {
  const wrappedFactory = () => {
    directiveInjectStart()
    def.factory()
    directiveInjectEnd()
  }
  return { factory: wrappedFactory }
}
```

So that compiler only generates `inject()` for DI types.

devversion replied:
Possibly. That looks reasonable to me at first glance. Especially since there already is logic to allow switching the `inject` implementation. 

I'm not sure what's the reasoning behind `directiveInject`, but I guess the separation of directive-level injection and module-level/root injection. Maybe @kara has some thoughts on this.

IgorMinar commented:
@alexeagle do we need this in version 8 or can it wait until version 9? can you please help to estimate the impact?

kara commented:
It would be good to get this in version 8, since it's stellated to ivy

devversion committed:
feat(core): add undecorated-base-class migration schematic

Introduces a new upgrade schematic that automatically decorates
"@Directive"/"@Component" base classes that use dependency injection
with the "@Directive" decorator. The new dummy directives are also
added to the corresponding "NgModule" declarations in order to not
cause any compilation warnings.

Adding the "@Directive" decorator ensures that "ngtsc" can properly
create the needed Ivy fields so that dependency injection works
as expected once Ivy is enabled in the application.

References FW-1238

devversion committed:
refactor(core): create google3 tslint rule for undecorated-base-class migration

Refactors the undecorated-base-class migration to support
a Google3 tslint rule that can be used for replacements or
just for enforcing decorated base-classes which are mandatory
in Ivy.

devversion commented:
Closing in favor of a new PR that will be opened soon.
